solidity/20241126-upgrade-path-testing.md

```diff
@@ -0,0 +1,119 @@
+3. run tests against the upgraded system.
+
+Put another way, we need two different methods for setting up the foundry tests which are
+created by inheriting from `CommonTest`.
+
+# Summary
+
+<!-- Most (if not all) documents should have a summary.
+While the length will likely be proportional to the length of the full document,
+the summary should be as succinct as possible. -->
+
+# Problem Statement + Context
+
+Our test suite is currently not able to replicate the system from a previous release, which
+prevents us from testing the upgrade path to the system currently under development.
+
+As we're moving towards onchain upgrades via the OPCM, we want to be able to test that:
+
+1. The upgrade works to move from the previous release to the system on `develop`.
+2. The upgraded system passes the same set of unit test as the freshly deployed system on `develop`.
+
+# Proposed Solution
+
+## Deploying Superchain contracts with `op-deployer`
+
+We propose to extend `op-deployer bootstrap` to enable deploying superchain contracts.
+
+```shell
+op-deployer bootstrap superchain <artifacts-locator>
+```
+
+## Setting up the system to test
+
+The current foundry testing uses the following high level call flow.
+
+```mermaid
+graph LR
+    A["CommonTest.setUp()"] --> B["Setup.L1()"] --> C["Deploy.run()"]
+    C --> D["DeployImplementations.run()"]
+    C --> E["DeploySuperchain.run()"]
+    C --> F["OPCM.deploy()"]
+```
+
+This is roughly what each component does:
+
+- **`CommonTest.setUp()`:**
+  - Defines system config settings (ie. `useInterop`).
+  - provides reusable addresses (`alice` and `bob`)
+  - provides reusable generic contracts (`ERC20`)
+- **`Setup.L1()`:**
+  - Reads addresses from the deployments file stored on disk
+  - labels addresses to make traces more readable
+  - ie.
+    ```solidity
+      optimismPortal = IOptimismPortal(deploy.mustGetAddress("OptimismPortalProxy"));
+      vm.label(address(optimismPortal), "OptimismPortal");
+    ```
+- **`Deploy.run()`:**
+  - Deploys all necessary contracts via calls to `DeploySuperchain`, `DeployImplementations`.
+- **`DeploySuperchain.run()`:**
+  - Deploys superchain contracts (`SuperchainConfig` and `ProtocolVersions`).
+- **`DeployImplementations.run()`:**
+  - Deploys contracts (implementations and singletons) necessary for upgrading to the system on
+    `develop`.
+- **`OPCM.deploy()`:**
+  - Deploys all proxies and bespoke singleton contracts as necessary for a new OP Chain.
+
+This work would replace the `Deploy` script with a new `Upgrade` script, resulting in the following
+call flow:
+
+```mermaid
+graph LR
+    A["CommonTest.setUp()"] --> B["Setup.L1()"] -->  C["Upgrade.run()"]
+    C --> D["DeployImplementations.run()"]
+    C --> E["OPCM.upgrade()"]
+```
+
+A description of what new components would do is:
+
+- **`Upgrade.run()`:**
+  - Calls `op-deployer bootstrap superchain` to deploy new superchain contracts (`SuperchainConfig`
```

Comment:
Is executing this command analogous to running both `DeploySuperchain.s.sol` and `DeployImplementations.s.sol`, but only having those scripts deploy contracts that changed from the last release? Don't need to get into implementation here, but one idea I've had for doing this is have those scripts use create2, precompute deploy addresses, and skip when code already exists. This requires a one-time redeploy of all contracts using this method, which we have to do for isthmus anyway since we're bumping to 0.8.25

Reply:
Ah, and `op-deployer bootstrap opcm` deploys the new implementations before deploying OPCM?